Added user model attributes for Github token api usage.

andrewjumanca opened the pull request:
- Added user GitHub authorization token with auto encryption/decryption
- Added test cases to cover response parsing and user model validation
- Split up from_omniauth function into smaller methods to reduce ABC complexity
- Fixed listing errors
Resolves issue #63

GitRepsOfficial commented on app/models/user.rb:
```diff
@@ -12,13 +12,24 @@ class User < ApplicationRecord
-      # If you are using confirmable and the provider(s) you use validate emails,
-      # uncomment the line below to skip the confirmation emails.
-      # user.skip_confirmation!
+      user.assign_email(auth)
```
This is a good example of how linting rules can guide you toward higher quality code. It doesn't give you an explicit answer but at least forces the conversation.

I like your instinct to separate out the logic into separate methods. That was a good first step, but getting around the naming convention by using a synonym is sort of a [code smell](https://en.wikipedia.org/wiki/Code_smell). It's not really your fault because we started with example code that lead us in this direction. So we should think about how to refactor this in a better way.

Let's zoom out and take a look at the larger system we're working with. In the [omniauth_callbacks_controller](https://github.com/GitRepsCommunity/platform/blob/c0a0fc4df82863adc8477b479abccc1df0764d37/app/controllers/users/omniauth_callbacks_controller.rb#L8), we are passing the raw object from the request (`request.env['omniauth.auth']`) into `from_omniauth`. We need to determine where the [separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns) should be between the `User` model and the OmniAuth response and controller.

The `User` model is primarily responsible for handling database ([CRUD](https://en.wikipedia.org/wiki/Create,_read,_update_and_delete)) operations for the user record. The `from_omniauth` method extends its responsibility and couples it to the structure of the OmniAuth response object. So the `User` model has to know about validating and parsing that object.

One alternative is to extract all of the OmniAuth logic from the `User` model and handle it somewhere else. Since the response is an object with attributes that need to be parsed and validated, it might be nice to define a new model.

Here's what it might look like:

Add `app/models/omniauth_response.rb`
```ruby
class OmniauthResponse
  include ActiveModel::Model

  def initialize(response:)
    @response = response
  end

  def provider
    @response['provider']
  end

  def uid
    @response['uid']
  end

  def email
    primary_email ? primary_email['email'] : @response['info']['email']
  end

  def primary_email
    @response['extra']['all_emails'].find { |e| e['primary'] == true }
  end

  def github_username
    @response['extra']['raw_info']['login']
  end

  def github_token
    @response['credentials']['token']
  end

  def profile_pic_url
    @response['extra']['raw_info']['avatar_url']
  end
end
```

Change `app/controllers/users/omniauth_callbacks_controller.rb`
```ruby
    def github
      auth = OmniauthResponse.new(response: request.env['omniauth.auth'])
      @user = User.from_omniauth(auth: auth)

      if @user.persisted?
        sign_in_and_redirect @user, event: :authentication
        set_flash_message(:notice, :success, kind: 'Github') if is_navigational_format?
      else
        session['devise.github_data'] = request.env['omniauth.auth'].except!(:extra)
        redirect_to new_user_registration_url
      end
    end
```

Change `app/models/user.rb`
```ruby
  def self.from_omniauth(auth)
    find_or_create_by(provider: auth.provider, uid: auth.uid) do |user|
      user.password = Devise.friendly_token[0, 20]
      user.email = auth.email
      user.github_username = auth.github_username
      user.github_token = auth.github_token
      user.profile_pic_url = auth.profile_pic_url
    end
  end
```

This allows us to keep all of the interfaces clean and simple. The new model uses [ActiveModel](https://edgeguides.rubyonrails.org/active_model_basics.html#model) which gives us the conveniences (validations, etc.) of a normal Rails model, but doesn't actually persist to the database.

andrewjumanca replied:
@GitRepsOfficial Makes sense! I'll look into the ActiveModel documentation for future reference. I'll have the updated changes ready today. Thanks for the explanations too, very insightful!

andrewjumanca commented:
@GitRepsOfficial I've fixed refactoring of from_omniauth, tests, and added omniauth_response, but still getting too high of an ABC ratio for the github function in the callbacks controller. Any ideas to fix this without just making a bunch of private functions again?

GitRepsOfficial commented:
> @GitRepsOfficial I've fixed refactoring of from_omniauth, tests, and added omniauth_response, but still getting too high of an ABC ratio for the github function in the callbacks controller. Any ideas to fix this without just making a bunch of private functions again?

Private functions are totally fine to use. It's more about having a clear separation of concerns. In this case, everything happening in the controller seems to make sense there.

I think something like this would work:
```ruby
module Users
  class OmniauthCallbacksController < Devise::OmniauthCallbacksController
    skip_before_action :verify_authenticity_token, only: :github

    def github
      @user = user_from_omniauth(request)

      if @user.persisted?
        sign_in_and_redirect @user, event: :authentication
        set_flash_message(:notice, :success, kind: 'Github') if is_navigational_format?
      else
        session['devise.github_data'] = request.env['omniauth.auth'].except!(:extra)
        redirect_to new_user_registration_url
      end
    end

    protected

    def user_from_omniauth(request)
      auth = OmniauthResponse.new(response: request.env['omniauth.auth'])
      User.from_omniauth(auth:)
    end
  end
end

```

GitRepsOfficial commented on test/models/user_test.rb:
```diff
@@ -3,7 +3,61 @@
+          'html_url' => 'https://github.com/andrewjumanca'
+        },
+        'all_emails' => [
+          { 'email' => '59379906+andrewjumanca@users.noreply.github.com' },
```
You'll probably want to replace your personal emails with fake ones, especially since this is a public repo. Something like `test@example.com` would work fine.

GitRepsOfficial commented on test/fixtures/omniauth_response.json:
```diff
@@ -0,0 +1,32 @@
+          "email": "12345678+exampleuser@users.noreply.github.com"
+        },
+        {
+          "email": "example@user.com"
```
"Fake" email addresses for testing can occasionally cause real world issues if you're not careful.

This actually happened at my job within the last couple of months. Our QA team was creating fake accounts to manually test user flows in the pre-prod environment. Our automated email system ended up spamming some random person who eventually contacted us to stop using their domain.

I only mention this because [user.com](https://www.user.com/) is a real site. Since we aren't sending email here it's not a big deal, but it's worth mentioning for future testing.

I like to use [example.com](https://example.com/) because it explicitly states:

>This domain is for use in illustrative examples in documents. You may use this domain in literature without prior coordination or asking for permission.

andrewjumanca replied:
Noted, that makes sense. Good to know for writing future testing data!
